### PR TITLE
Feature/202 skills endpoint

### PR DESCRIPTION
**Pull request by gustavo-rojas-jalafund**
[AB#18](https://dev.azure.com/funjaladev/7867a1c5-5ac7-4c7f-baff-87d740bf4fc1/_workitems/edit/18) https://dev.azure.com/funjaladev/DevLevel3/_workitems/edit/202

# Overview

Implement Skills controller
Implement Skills service
Implement Skills DAL that returns fake data from memory
Implement unit tests for API layer
Implement unit tests for service layer

### CODE REVIEW

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        private readonly ISkillsService skillsService;
+
+        public GetSkillsByName(ISkillsService skillsService)
+        {
+            this.skillsService = skillsService;
+        }
+
+        [FunctionName("GetSkillsByName")]
+        [OpenApiOperation(operationId: "GetSkillsByName", tags: new[] { "Skills" })]
+        [OpenApiParameter(name: "name", In = ParameterLocation.Query, Required = false, Type = typeof(string), Description = "The resume identifier.")]
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs]:
```suggestion
        [OpenApiParameter(name: "name", In = ParameterLocation.Query, Required = false, Type = typeof(string), Description = "The name of the skill to search by.")]
```

File: src/api/Jalasoft.TeamUp.Resumes.API/Startup.cs [hunk 2]
```diff
@@ -19,11 +19,9 @@ public override void Configure(IFunctionsHostBuilder builder)
             builder.Services.AddScoped<IHealthService, HealthService>();
             builder.Services.AddScoped<IHealthRepository, HealthRepository>();
             builder.Services.AddScoped<IResumesService, ResumesService>();
-#if DEBUG
-            builder.Services.AddScoped<IRepository<Resume>, ResumeSQLRepository>();
-#else
             builder.Services.AddScoped<IRepository<Resume>, ResumesInMemoryRepository>();
-#endif
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/Startup.cs:26]: I think you shouldn't make changes to this file, please double check

File: src/models/Jalasoft.TeamUp.Resumes.Models/Skill.cs [hunk 3]
```diff
@@ -6,6 +6,6 @@ public class Skill
     {
         public Guid Id { get; set; }
```
Comment (cristian-chavez-jalafund) [inline src/models/Jalasoft.TeamUp.Resumes.Models/Skill.cs]: so far I understand, this should be string, right @ela-ocana-jalafund?

File: src/api/Jalasoft.TeamUp.Resumes.API/Startup.cs (hunk 2 above)
Comment (cristian-chavez-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Startup.cs:26]: or at least these lines shouldn't be affected

File: src/models/Jalasoft.TeamUp.Resumes.Models/Skill.cs (hunk 3 above)
Comment (ela-ocana-jalafund) [reply src/models/Jalasoft.TeamUp.Resumes.Models/Skill.cs]: Right. We need to change that.

Comment (gustavo-rojas-jalafund) [reply src/models/Jalasoft.TeamUp.Resumes.Models/Skill.cs]: Yes, Ela told me, it is already corrected

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs (hunk 1 above)
Comment (gustavo-rojas-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs]: Done

File: src/api/Jalasoft.TeamUp.Resumes.API/Startup.cs (hunk 2 above)
Comment (gustavo-rojas-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Startup.cs:26]: Done

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        [OpenApiParameter(name: "name", In = ParameterLocation.Query, Required = false, Type = typeof(string), Description = "The name of the skill to search by.")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.OK, contentType: "application/json", bodyType: typeof(Skill[]), Description = "Successful response")]
+        [OpenApiResponseWithoutBody(statusCode: HttpStatusCode.NotFound, Description = "Resource not found")]
+
+        public IActionResult Run(
+        [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "v1/skills")] HttpRequest req)
+        {
+            req.Query.TryGetValue("name", out StringValues name);
+            var resumes = this.skillsService.GetSkills(name);
+            return new OkObjectResult(resumes);
```
Comment (ela-ocana-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs:35]: Could you change resumes by skills? Line 34 and line 35. I believe we didn't see that.

File: src/core/Jalasoft.TeamUp.Resumes.Core/SkillsService.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        private readonly ISkillsRepository skillsRepository;
+
+        public SkillsService(ISkillsRepository skillsRepository)
+        {
+            this.skillsRepository = skillsRepository;
+        }
+
+        public Skill[] GetSkills(string name)
+        {
+            return (Skill[])this.skillsRepository.GetSkills(name);
```
Comment (rodrigo-baldiviezo-jalafund) [inline src/core/Jalasoft.TeamUp.Resumes.Core/SkillsService.cs]: It's necessary to cast?. Your ISkillsRepository in GetSkills method return Skill[]

**Review by rodrigo-baldiviezo-jalafund — COMMENTED** [review]
Your [US](https://dev.azure.com/funjaladev/DevLevel3/_boards/board/t/DevLevel3%20Team/Stories/?workitem=202) has this AC:
- Implement unit tests for API layer
- Implement unit tests for service layer

I don't see the tests files for skills

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs (hunk 4 above)
Comment (gustavo-rojas-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs]: Done

File: src/core/Jalasoft.TeamUp.Resumes.Core/SkillsService.cs (hunk 5 above)
Comment (gustavo-rojas-jalafund) [reply src/core/Jalasoft.TeamUp.Resumes.Core/SkillsService.cs]: Done

**Review by ela-ocana-jalafund — APPROVED** [review]

**Review by freddy-perez-jalafund — APPROVED** [review]
Approved

**Review by rodrigo-baldiviezo-jalafund — APPROVED** [review]
LGTM

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+        public GetSkillsByName(ISkillsService skillsService)
+        {
+            this.skillsService = skillsService;
+        }
+
+        [FunctionName("GetSkillsByName")]
+        [OpenApiOperation(operationId: "GetSkillsByName", tags: new[] { "Skills" })]
+        [OpenApiParameter(name: "name", In = ParameterLocation.Query, Required = false, Type = typeof(string), Description = "The name of the skill to search by.")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.OK, contentType: "application/json", bodyType: typeof(Skill), Description = "Successful response")]
```
Comment (Raul-Gamarra-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs:27]: `bodyType: typeof(Skill),`
this should be type of an array of skills

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        [OpenApiResponseWithoutBody(statusCode: HttpStatusCode.NotFound, Description = "Resource not found")]
+
+        public IActionResult Run(
+        [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "v1/skills")] HttpRequest req)
+        {
+            req.Query.TryGetValue("name", out StringValues name);
```
Comment (Raul-Gamarra-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs:33]: add exception handling with the custom exception created for this project

File: src/core/Jalasoft.TeamUp.Resumes.Core/SkillsService.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+        private readonly ISkillsRepository skillsRepository;
+
+        public SkillsService(ISkillsRepository skillsRepository)
+        {
+            this.skillsRepository = skillsRepository;
+        }
+
+        public Skill[] GetSkills(string name)
+        {
+            return this.skillsRepository.GetSkills(name).ToArray();
```
Comment (Raul-Gamarra-jalafund) [inline src/core/Jalasoft.TeamUp.Resumes.Core/SkillsService.cs:21]: validate that this result is not null and throw an exception when it is

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/GetSkillsByNameTests.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            this.mockService.Setup(service => service.GetSkills(".NET")).Returns(new Skill[2]);
+
+            // Act
+            var response = this.getSkillsByName.Run(request);
+
+            // Assert
+            var okObjectResult = Assert.IsType<OkObjectResult>(response);
+            Assert.IsType<Skill[]>(okObjectResult.Value);
+        }
+    }
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Resumes.API.Tests/GetSkillsByNameTests.cs:43]: we need a new test here for not found response

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetSkillsByNameCoreTests.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            var stubSkillList = GetTestSkills();
+            this.mockRepository.Setup(repository => repository.GetSkills(".NET")).Returns(stubSkillList);
+
+            // Act
+            var result = this.skillsService.GetSkills(".NET");
+
+            // Assert
+            Assert.Equal(3, result.Length);
+        }
+    }
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetSkillsByNameCoreTests.cs:58]: we need a new test here for not found response

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs (hunk 7 above)
Comment (Raul-Gamarra-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs:33]: reach out to @paola-quintanilla-jalafund

File: src/core/Jalasoft.TeamUp.Resumes.Core/SkillsService.cs (hunk 8 above)
Comment (Raul-Gamarra-jalafund) [reply src/core/Jalasoft.TeamUp.Resumes.Core/SkillsService.cs:21]: reach out to @paola-quintanilla-jalafund

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs (hunk 6 above)
Comment (gustavo-rojas-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetSkillsByName.cs:27]: Done

**Review by cristian-chavez-jalafund — APPROVED** [review]

**Pull request merged**